localstack/services/s3/provider.py

```diff
@@ -1,12 +1,252 @@
+        request: ListObjectsRequest,
+    ) -> ListObjectsOutput:
+        bucket_name = request.get("Bucket", "")
+        response = call_moto_with_exception_patching(context, bucket_name)
+
+        if "Marker" not in response:
+            response["Marker"] = request.get("Marker") or ""
+
+        encoding_type = request.get("EncodingType")
+        if "EncodingType" not in response and encoding_type:
+            response["EncodingType"] = encoding_type
+
+        # fix URL-encoding of Delimiter
+        if delimiter := response.get("Delimiter"):
+            delimiter = delimiter.strip()
+            if delimiter != "/":
+                response["Delimiter"] = quote(delimiter)
+
+        if "BucketRegion" not in response:
+            moto_backend = get_moto_s3_backend(context)
+            bucket = get_bucket_from_moto(moto_backend, bucket=bucket_name)
+            response["BucketRegion"] = bucket.region_name
+
+        return ListObjectsOutput(**response)
+
+    @handler("ListObjectsV2", expand=False)
+    def list_objects_v2(
+        self,
+        context: RequestContext,
+        request: ListObjectsV2Request,
+    ) -> ListObjectsV2Output:
+        bucket_name = request.get("Bucket", "")
+        response = call_moto_with_exception_patching(context, bucket_name)
+
+        encoding_type = request.get("EncodingType")
+        if "EncodingType" not in response and encoding_type:
+            response["EncodingType"] = encoding_type
+
+        # fix URL-encoding of Delimiter
+        if delimiter := response.get("Delimiter"):
+            delimiter = delimiter.strip()
+            if delimiter != "/":
+                response["Delimiter"] = quote(delimiter)
+
+        if "BucketRegion" not in response:
+            moto_backend = get_moto_s3_backend(context)
+            bucket = get_bucket_from_moto(moto_backend, bucket=bucket_name)
+            response["BucketRegion"] = bucket.region_name
+
+        return ListObjectsV2Output(**response)
+
+    @handler("HeadObject", expand=False)
+    def head_object(
+        self,
+        context: RequestContext,
+        request: HeadObjectRequest,
+    ) -> HeadObjectOutput:
+        bucket_name = request.get("Bucket", "")
+        response = call_moto_with_exception_patching(context, bucket_name)
+
+        return HeadObjectOutput(**response, AcceptRanges="bytes")
+
+    @handler("GetObject", expand=False)
+    def get_object(self, context: RequestContext, request: GetObjectRequest) -> GetObjectOutput:
+        response = call_moto_with_exception_patching(context, bucket=request.get("Bucket", ""))
+        return GetObjectOutput(**response, AcceptRanges="bytes")
+
+    @handler("PutObject", expand=False)
+    def put_object(
+        self,
+        context: RequestContext,
+        request: PutObjectRequest,
+    ) -> PutObjectOutput:
+        if checksum_algorithm := request.get("ChecksumAlgorithm"):
+            verify_checksum(checksum_algorithm, context.request.data, request)
```

Comment:
Just as a reference for a potential implementation as a handler (see other comment):
In this case, the checksum seems to be optional, but in other cases it seems like the checksum has to be enforced (f.e. `PutBucketAcl`).